File: 2015/ruby/day14/part2.rb
#!/usr/bin/ruby

begin
    File.open('input.txt', 'r') do |file|
        reindeer = []

        file.each_line do |line|
            sentence = line.split(' ')
            reindeer.push({
                'speed' => sentence[3].to_i,
                'flyTime' => sentence[6].to_i,
                'restTime' => sentence[13].to_i,
                'distance' => 0,
                'flying' => true,
                'timeLeft' => sentence[6].to_i,
                'score' => 0
            })
        end

        for i in 1..2503
            reindeer.each do |stats|
                stats['timeLeft'] -= 1

                if stats['flying']
                    stats['distance'] += stats['speed']

                    if stats['timeLeft'] == 0
                        stats['flying'] = false
                        stats['timeLeft'] = stats['restTime']
                    end

                    next
                end

                if stats['timeLeft'] == 0
                    stats['flying'] = true
                    stats['timeLeft'] = stats['flyTime']
                end
            end

            reindeer.group_by{ |value| value['distance'] }.max.last.each do |stats|
                stats['score'] += 1
            end
        end

        puts reindeer.max_by{ |value| value['score'] }['score']
    end
rescue SystemCallError
    puts 'input.txt not found.'
end

File: 2016/python/day20/part2.py
#!/usr/bin/python

with open('input.txt') as file_handler:
    ranges = [map(int, line.rstrip().split('-')) for line in file_handler]

ip = 0
allowed = 0

while ip <= 4294967295:
    upper = next((rng[1] for rng in ranges if rng[0] <= ip <= rng[1]), None)

    if upper:
        ip = upper + 1
        continue

    allowed += 1
    ip += 1

print allowed
